### DISCUSSION TITLE
Filtering and sorting on name

### DISCUSSION THREAD

**Discussion by Mouwrice** [SELab-2/OSOC-2#287] [General]
while trying to implement the filter for searching and sorting on the username I've encountered following problem.

We want to search and sort on name. however, we have this stored in 2 fields: firstname and lastname. Prisma doesn't allow us to search on these fields combined. The 2 solutions we have are:

 - combine the first- and lastname field into 1 field named name, and rewrite all the code that is using this currently. This would result in a good filter/search mechanism.
 -  split the filter to search/sort on first and lastname separately. This way we don't have to rewrite other code, BUT the filter/search will be more annoying to use. Furthermore would this keep giving some problems because with github login there is no lastname.

⚠️ We should decide this ASAP ⚠️

**Note:** Moved #285 to this discussion.

**Comment by Mouwrice** [comment]
[jay-tux](https://github.com/jay-tux)(Jay/Jonas) commented [1 hour ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091392889)

Why don't we just pull all names from the db, then filter using toUpperCase and substring in the BE code itself? As far as I know, there has been no filtering code yet written.

**Reply by Mouwrice** [reply]
[Mouwrice](https://github.com/Mouwrice) commented [1 hour ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091394659) •

We want to avoid filtering in the backend as this is very inefficient. Databases are specifically designed for this purpose.

**Reply by Mouwrice** [reply]
[BramDevlaminck](https://github.com/BramDevlaminck)(Bram Devlaminck) commented [1 hour ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091395200) •

There is no code yet indeed, but it would be a lot cleaner to filter in the database itself, furthermore have Norick and I found a clean syntax to filter in there, the code would be a lot more readable

**Comment by Mouwrice** [comment]
[Mouwrice](https://github.com/Mouwrice) commented [1 hour ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091393202)

Since GitHub OAuth doesn't has the ability to get the firstname and lastname it seems easier to just remove the lastname field altogether and rewrite the code (which would be annoying be simple to do). And it will make the filter usage much more pleasant.

**Comment by Mouwrice** [comment]
[HuanYu-Tan](https://github.com/HuanYu-Tan) commented [1 hour ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091407888) •

I think that removing the lastname is the easiest solution.

**Comment by Mouwrice** [comment]
[BramDevlaminck](https://github.com/BramDevlaminck)(Bram Devlaminck) commented [1 hour ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091418891)

There is another (very ugly) solution: we could also keep another field name that is just a joined version of firstname and lastname. but I don't think that this is something we want to do (duplicate data,...)

**Reply by Mouwrice** [reply]
[nbeteram](https://github.com/nbeteram) commented [1 hour ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091421938) •

@BramDevlaminck that is indeed not the best option, I think.

**Comment by Mouwrice** [comment]
[nbeteram](https://github.com/nbeteram) commented [1 hour ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091421157)

I think splitting the filters is also a good option. That way, the users have different options and if they need the first name for something, it is easy to get it.

**Reply by Mouwrice** [reply]
[Mouwrice](https://github.com/Mouwrice) commented [39 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091434973)

The problem is, is that lastname is empty when using GitHub login. And the whole name is then in firstname, which would make searching very inconsistent and confusing.

**Comment by Mouwrice** [comment]
[KoenDesplenter](https://github.com/KoenDesplenter) commented [32 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091442082)

First of all, where would this filter be used?
If I'am not mistaken, student always have a first and lastname. And since the tool is really focused on the student part, filtering on both first and lastname is a huge benefit.
This would mean that person entries with a github account can't be filtered in the same way. But do we need login user filtering anyway?

**Reply by Mouwrice** [reply]
[BramDevlaminck](https://github.com/BramDevlaminck)(Bram Devlaminck) commented [28 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091446295) •

this filter would be used in the manage-loginUser screen where you can disable an account, set the account to admin,... see this issue

**Reply by Mouwrice** [reply]
[](https://github.com/KoenDesplenter)@[KoenDesplenter](https://github.com/KoenDesplenter)
KoenDesplenter commented [24 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091454500)

Since the amount login user will be limited in conparison to the amount of students, I would optimize the filter for the students.
I think we can overcome this filter problem by for example replacing null lastnames with an empty string. This way filtering is still possible however they would end up in front of real lastnames.

**Reply by Mouwrice** [reply]
[BramDevlaminck](https://github.com/BramDevlaminck)(Bram Devlaminck) commented [22 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091458510)

that does make sense, idk what the others think of this?

**Comment by Mouwrice** [comment]
[](https://github.com/Mouwrice)@[Mouwrice](https://github.com/Mouwrice)
Mouwrice commented [24 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091453642) •

For students there is indeed no issue with the current setup, we would be able to filter and sort on first name and last name.
And then for login users it doesn't seem that very important to filter on both first name and last name so then maybe we should just store the entire user name in the firstname field. And have the filter just set to name

Plus, the users first name and lastname will never be shown seperately. So there is not really a need here for seperating the full name.

**Reply by Mouwrice** [reply]
[BramDevlaminck](https://github.com/BramDevlaminck)(Bram Devlaminck) commented [13 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091479358) •

this would mean it's stored something like this?
|firstname|lastname|
|---------|---------|
|Bram Devlaminck | Devlaminck |

I wouldn't do this, this will result is some weird and unexpected behaviour

**Reply by Mouwrice** [reply]
[Mouwrice](https://github.com/Mouwrice) commented [12 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091483723) •

|firstname|lastname|
|---------|---------|
|Bram Devlaminck | "" |

No, the lastname for login users would always be empty (which is nonetheless already the case for users with github login)
And it would make every login user's information conform.

**Reply by Mouwrice** [reply]
[BramDevlaminck](https://github.com/BramDevlaminck)(Bram Devlaminck) commented [11 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091486164) •

and what would you do when there is a student-coach? idk if this student can log in?

**Reply by Mouwrice** [reply]
[](https://github.com/Mouwrice)@[Mouwrice](https://github.com/Mouwrice)
Mouwrice commented [9 minutes ago](https://github.com/SELab-2/OSOC-2/issues/285#issuecomment-1091489312)

We don't have any student-coach roles for login users. I dont think they will have an account, as they only help the other students on the project they're assigned to

**Comment by KoenDesplenter** [comment]
A quick summary of the solution, please correct me if I'm wrong
The datafields firstname and lastname will remain, for the purpose of filtering students on both fields.
A login user that uses email as login method, will have both first and lastname residing in the firstname field.
The benefit of this method being that login users with github had no lastname anyway so filtering is uniform across all login users.

**Reply by Mouwrice** [reply]
Splendid!

**Comment by BramDevlaminck** [comment]
Do we all agree on this solution and should we rename the field `firstname`?

**Reply by KoenDesplenter** [reply]
I wouldn't go through all the hassle to rename the field

**Reply by Mouwrice** [reply]
Let's just leave it. We already have enough work to do.
